### PR TITLE
[UXIT-1873] - Read from CMS config instead of YAML repositories

### PR DESCRIPTION
**Pull request by CharlyMartin**
## 📝 Description

This PR updates the category and subcategory Zod schemas to pull data from the CMS config instead of the YAML folder.

It also removes `getCategoryDataFromDirectory` and updates all code that relied on it to use `getEcosystemCMSCategories`, which reads from the CMS config.

## 🛠️ Key Changes

1. Updates category and subcategory Zod schemas to match the CMS format - https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/906/commits/7c92bd393242f506f7f1e95af5817dcd90c262d9
2. Removes `getCategoryDataFromDirectory` and introduces `getEcosystemCMSCategories` to retrieve category information - https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/906/commits/2ec38e30512a13e1f9d64cb98f9e2426b20ecf86
3. Introduces `mapCMSOptionsToListboxFormat` to format CMS options to the `Listbox` format - https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/906/commits/d95bf1384a202931021873815ae4e8aff00fe0db

### CODE REVIEW

**Review by mirhamasala — COMMENTED** [review]
@CharlyMartin - [This effort](https://www.notion.so/filecoin/FF-Ecosystem-Explorer-Rework-Categories-Subcategories-structure-14b7631f2825807e9ed6d87fef734eb9?pvs=4) should work in a similar way as the region implementation we did for Events.

This PR should be merged when everything is in place, that is, when the config has a single category field with just string values and the hierarchy is handled on the front. Let me know if that makes sense.

Thank you. 🙏🏼

(I did do a first review - Looking good!)

**Comment by mirhamasala** [comment]
@CharlyMartin - Yes, please!

>So, I’ll create a feature branch for this effort, and from there create smaller PRs. Once everything is ready, we’ll merge the feature branch into main and everything will be in place in one go. Is that what you had in mind?

**Comment by CharlyMartin** [comment]
This PR also tackles https://www.notion.so/filecoin/FF-Ecosystem-Explorer-Consolidate-getCategoriesFromDirectory-and-getCategoryDataFromDirector-1577631f28258042a0f6faa73b34acc4?pvs=4

File: src/app/_utils/categoryUtils.ts [hunk 1]
```diff
@@ -1,26 +1,13 @@
-import { type CategoryMap, type CategoryYAMLData } from '@/types/categoryTypes'
+import { type CategoryMap } from '@/types/categoryTypes'
```
Comment (mirhamasala) [inline src/app/_utils/categoryUtils.ts:1]: Mega nit.

```suggestion
import type { CategoryMap } from '@/types/categoryTypes'
```

File: src/app/_utils/categoryUtils.ts [hunk 2]
```diff
@@ -1,26 +1,13 @@
...
 import { type CMSCollectionName } from '@/types/cmsConfig'
```
Comment (mirhamasala) [inline src/app/_utils/categoryUtils.ts:2]: I thought we had something enforcing this.

```suggestion
import type { CMSCollectionName } from '@/types/cmsConfig'
```

File: src/app/_utils/getCMSFieldOptionsAndValidIds.ts [hunk 3]
```diff
@@ -1,6 +1,7 @@
-import { type CMSCollectionName, type CMSFieldOption } from '@/types/cmsConfig'
+import { type CMSCollectionName } from '@/types/cmsConfig'
```
Comment (mirhamasala) [inline src/app/_utils/getCMSFieldOptionsAndValidIds.ts:1]:
```suggestion
import type { CMSCollectionName } from '@/types/cmsConfig'
```

File: src/app/ecosystem-explorer/hooks/useEcosystemCategory.ts
Comment (mirhamasala) [inline src/app/ecosystem-explorer/hooks/useEcosystemCategory.ts:1]: Nothing to do with this for now - Might add an follow up work to refactor this.

```
import { useMemo } from 'react'
import type { NextServerSearchParams } from '@/types/searchParams'
import { CATEGORY_KEY } from '@/constants/searchParams'
import type { EcosystemProject } from '../types/ecosystemProjectType'
import { getEcosystemCMSCategories } from '../utils/getEcosystemCMSCategories'

// Types
type CategoryTreeItem = {
  value: string
  label: string
  subcategories: Array<SubcategoryWithCount>
}

type SubcategoryWithCount = {
  value: string
  label: string
  parent_category: string
  count: number
}

type UseEcosystemCategoryProps<Entry extends EcosystemProject> = {
  searchParams: NextServerSearchParams
  categories: ReturnType<typeof getEcosystemCMSCategories>['categories']
  subcategories: ReturnType<typeof getEcosystemCMSCategories>['subcategories']
  entries: Array<Entry>
}

type UseEcosystemCategoryResult<Entry> = {
  filteredEntries: Array<Entry>
  categoryTree: Array<CategoryTreeItem>
  categoryQueries: string[] | null
}

// Helper functions
function getEntriesPerSubcategory<Entry extends EcosystemProject>(
  entries: Array<Entry>
): Map<string, number> {
  const counts = new Map()
  
  entries.forEach((entry) => {
    const subcategory = entry.subcategories[0]
    counts.set(subcategory, (counts.get(subcategory) || 0) + 1)
  })
  
  return counts
}

function buildCategoryTree(
  categories: UseEcosystemCategoryProps<any>['categories'],
  subcategories: UseEcosystemCategoryProps<any>['subcategories'],
  entryCounts: Map<string, number>
): Array<CategoryTreeItem> {
  return categories.map((category) => ({
    value: category.value,
    label: category.label,
    subcategories: subcategories
      .filter((subcategory) => subcategory.parent_category === category.value)
      .map((subcategory) => ({
        ...subcategory,
        count: entryCounts.get(subcategory.value) || 0,
      })),
  }))
}

function filterEntriesByCategory<Entry extends EcosystemProject>(
  entries: Array<Entry>,
  categoryIds: string[] | null
): Array<Entry> {
  if (!categoryIds) return entries
  return entries.filter((entry) => categoryIds.includes(entry.subcategories[0]))
}

// Main hook
export function useEcosystemCategory<Entry extends EcosystemProject>({
  searchParams,
  entries,
  categories,
  subcategories,
}: UseEcosystemCategoryProps<Entry>): UseEcosystemCategoryResult<Entry> {
  const categoryParams = searchParams[CATEGORY_KEY]
  const categoryIds = parseCategoryQueryParam(categoryParams)

  const entriesPerSubcategory = useMemo(
    () => getEntriesPerSubcategory(entries),
    [entries]
  )

  const categoryTree = useMemo(
    () => buildCategoryTree(categories, subcategories, entriesPerSubcategory),
    [categories, subcategories, entriesPerSubcategory]
  )

  const filteredEntries = useMemo(
    () => filterEntriesByCategory(entries, categoryIds),
    [entries, categoryIds]
  )

  return {
    filteredEntries,
    categoryTree,
    categoryQueries: categoryIds,
  }
}
```

Comment (mirhamasala) [inline src/app/ecosystem-explorer/hooks/useEcosystemCategory.ts:1]:
```ts
categories.map(({ value, label }) => ({
        value,
        label,
        subcategories: subcategories
          .filter(({ parent_category }) => parent_category === value)
```

File: src/app/ecosystem-explorer/page.tsx [hunk 4]
```diff
@@ -161,15 +161,15 @@ export default function EcosystemExplorer({ searchParams }: Props) {
                       const isFirstTwoImages = i < 2
                       const category = findOrThrow(
                         subcategories,
-                        (category) => category.slug === categoryId,
+                        (category) => category.value === categoryId,
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/page.tsx:164]:
```suggestion
                        ({value) => value === categoryId,
```

**Review by mirhamasala — CHANGES_REQUESTED** [review]
Still a little confused about the order of things and the bigger picture - But trusting the process. 🔮🦀

Mostly nits - mostly Scout.

We should remove `readAndValidateYamlFiles` at this point, right?

Thank you! 🙏🏼

File: src/app/ecosystem-explorer/[slug]/page.tsx [hunk 5]
```diff
@@ -53,6 +57,11 @@ export default function EcosystemProject({ params }: EcosystemProjectProps) {
     subcategories,
   } = data
 
+  const projectSubcategory = findOrThrow(
+    cmsSubcategories,
+    (cmsSubcategory) => cmsSubcategory.value === subcategories[0],
+  )
```
Comment (mirhamasala) [inline src/app/ecosystem-explorer/[slug]/page.tsx:63]:
```suggestion
const projectSubcategory = findOrThrow(
  cmsSubcategories,
  ({ value }) => value === subcategories[0]
)
```

File: src/app/ecosystem-explorer/schemas/CategorySchemas.ts [hunk 6]
```diff
@@ -1,41 +1,31 @@
...
-  slug: SubcategorySlug,
-  name: z.string(),
-  parent_category: CategorySlug,
+export const SubcategoryOptionSchema = z.object({
+  value: SubcategoryValue,
+  label: z.string(),
+  parent_category: CategoryValue,
 })
+
+function createSchemaForField(fieldName: 'category' | 'subcategory') {
```
[the remark this answers is not in the document]
Comment (CharlyMartin) [reply src/app/ecosystem-explorer/schemas/CategorySchemas.ts:24]: `createSchemaForField` creates a Zod schema for a given CMS field
`getCMSOptionsForField` returns an array of CMS options

[the remark this answers is not in the document]
Comment (mirhamasala) [reply src/app/ecosystem-explorer/schemas/CategorySchemas.ts:24]: They both use part of the same code ad defined in `getCMSOptionsForField`..

[the remark this answers is not in the document]
Comment (CharlyMartin) [reply src/app/ecosystem-explorer/schemas/CategorySchemas.ts:24]: I've put them in separate files

[the remark this answers is not in the document]
Comment (mirhamasala) [reply src/app/ecosystem-explorer/schemas/CategorySchemas.ts:24]: How do you mean?

File: src/app/_utils/categoryUtils.ts (hunk 1 above)
Comment (CharlyMartin) [reply src/app/_utils/categoryUtils.ts:1]: No, you're right!

File: src/app/ecosystem-explorer/schemas/CategorySchemas.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (CharlyMartin) [reply src/app/ecosystem-explorer/schemas/CategorySchemas.ts]: https://github.com/FilecoinFoundationWeb/filecoin-foundation/pull/906/commits/f29fdfe5280e0af2fea8efa665f2bed760c93f2a

File: src/app/ecosystem-explorer/hooks/useEcosystemCategory.ts
Comment (CharlyMartin) [reply src/app/ecosystem-explorer/hooks/useEcosystemCategory.ts:1]: This hook will be broken down in the next PR. The filtering will most likely be performed in `useFilter`.

Then, we can see how we can improve the rest. Does that work?

Comment (CharlyMartin) [reply src/app/ecosystem-explorer/hooks/useEcosystemCategory.ts:1]: I like it :)

Let's do it in the next PR where `subcategories` becomes `subcategory`!

File: src/app/ecosystem-explorer/schemas/CategorySchemas.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mirhamasala) [reply src/app/ecosystem-explorer/schemas/CategorySchemas.ts]: NAISE.

**Comment by CharlyMartin** [comment]
Thank you for the review @mirhamasala 

> We should remove `readAndValidateYamlFiles` at this point, right?

Definitely! Good catch!

> Mostly nits - mostly Scout.

I'm gonna try to implement some of them in the following PR :)

**Pull request merged**